packages/style-engine/class-wp-style-engine-renderer.php

```diff
@@ -0,0 +1,194 @@
+<?php
+/**
+ * WP_Style_Engine_Renderer
+ *
+ * A library of CSS rule generators and render functions.
+ *
+ * @TODO Consider splitting out the rule generators (?).
+ * Or creating a base interface and then a bunch of separate render classes for each "layer", e.g.,
+ * one for block supports, one for global styles.
+ *
+ * @package Gutenberg
+ */
+
+if ( class_exists( 'WP_Style_Engine_Renderer' ) ) {
+	return;
+}
+
+/**
+ * Renders CSS on the frontend.
+ *
+ * @access private
+ */
+class WP_Style_Engine_Renderer {
+	/**
+	 * Prints registered styles in the page head or footer.
+	 *
+	 * @TODO this shares code with the styles engine class in generate(). Centralize.
+	 *
+	 * @see $this->enqueue_block_support_styles
+	 */
+	public static function render_registered_block_supports_styles() {
+		$style_engine         = WP_Style_Engine::get_instance();
+		$block_support_styles = $style_engine->get_registered_styles();
+
+		if ( empty( $block_support_styles ) ) {
+			return;
+		}
+
+		$output = '';
+
+		foreach ( $block_support_styles as $selector => $css_definitions ) {
+			$output .= self::generate_css_rule( $selector, $css_definitions, array( 'prettify' => true ) );
+		}
+
+		echo "<style>\n$output</style>\n";
+	}
+
+	/**
+	 * Filters incoming CSS properties against WordPress Core's allowed CSS attributes in wp-includes/kses.php.
+	 *
+	 * @param string $property_declaration A CSS property declaration, e.g., `color: 'pink'`.
+	 *
+	 * @return string A filtered CSS property. Empty if not allowed.
+	 */
+	public static function sanitize_property_declaration( $property_declaration ) {
+		return esc_html( safecss_filter_attr( $property_declaration ) );
+	}
+
+	/**
+	 * Creates a string consisting of CSS property declarations suitable for the value of an HTML element's style attribute.
+	 *
+	 * @param array $css_definitions An collection of CSS definitions `[ [ 'color' => 'red' ] ]`.
```

Comment:
References to `$css_definitions` should rather be `$css_declarations` since they contain `property: value` pairs